src/main/java/com/svydovets/bibirnate/cache/Cache.java

```diff
@@ -0,0 +1,164 @@
+        lastClean = LocalDateTime.now();
+
+        cacheMap = new ConcurrentHashMap<>(CACHE_SIZE);
+        initializeKeyExtractorCommandMap();
+        initializeInvalidationCommandMap();
+    }
+
+    public void put(Key<?> key, Object value) {
+        Objects.requireNonNull(key, String.format(PARAMETER_CANNOT_BE_NULL, key));
+        Objects.requireNonNull(value, String.format(PARAMETER_CANNOT_BE_NULL, value));
+
+        cacheMap.put(key, value);
+
+        autoClean();
+    }
+
+    public Optional<Object> get(AbstractKeyParam<?> abstractKeyParam, Class<? extends KeyExtractorCommand> commandType) {
+        Objects.requireNonNull(abstractKeyParam, String.format(PARAMETER_CANNOT_BE_NULL, abstractKeyParam));
+        Objects.requireNonNull(commandType, String.format(PARAMETER_CANNOT_BE_NULL, commandType));
+
+        Optional<Key<?>> key = getKey(abstractKeyParam, commandType);
+
+        return key.map(cacheMap::get);
+    }
+
+    public void delete(AbstractKeyParam<?> abstractKeyParam, Class<? extends KeyExtractorCommand> commandType) {
+        Objects.requireNonNull(abstractKeyParam, String.format(PARAMETER_CANNOT_BE_NULL, abstractKeyParam));
+        Objects.requireNonNull(commandType, String.format(PARAMETER_CANNOT_BE_NULL, commandType));
+
+        getKey(abstractKeyParam, commandType).ifPresent(cacheMap::remove);
+    }
+
+    public void invalidateRelated(AbstractKeyParam<?> abstractKeyParam,
+                                  Class<? extends KeyExtractorCommand> extractCommandType,
+                                  Class<? extends InvalidationCommand> invalidationCommandType){
+        Objects.requireNonNull(abstractKeyParam, String.format(PARAMETER_CANNOT_BE_NULL, abstractKeyParam));
+        Objects.requireNonNull(extractCommandType, String.format(PARAMETER_CANNOT_BE_NULL, extractCommandType));
+
+
+        getKey(abstractKeyParam, extractCommandType)
+                .ifPresent(value -> invalidationCommandMap.get(invalidationCommandType)
+                        .executeInvalidate(cacheMap, value));
+    }
+
+    /**
+     * Makes clear for {@link Cache#cacheMap}.
+     */
+    public void clear() {
+        cacheMap.clear();
+    }
+
+    public void addKeyExtractorCommand(KeyExtractorCommand keyExtractorCommand) {
+        Objects.requireNonNull(keyExtractorCommand, String.format(PARAMETER_CANNOT_BE_NULL, keyExtractorCommand));
+
+        keyExtractorCommandMap.put(keyExtractorCommand.getClass(), keyExtractorCommand);
+    }
+
+    private void invalidateRelated(Key<?> key){
+
+    }
+
+    private void autoClean() {
+        if ((Float.valueOf(cacheMap.size()) / Float.valueOf(CACHE_SIZE)) * ONE_HUNDRED > MAX_CACHE_SIZE_PERCENTAGE) {
+            if (Duration.between(lastClean, LocalDateTime.now()).toHours() <= TWO_HOURS) {
+                throw new CacheOverloadException("Biberante Cache is overloaded! " +
+                        "Please check your settings and review if second level cache is enabled or what kind of " +
```

Comment:
How second lvl cache influences the first lvl cache?

Reply:
they are not related to each other. It's just a realization of LRU cache. So, you can create Cache on session level (1st lvl), and on sessionFactory lvl and take a control of both of them where u use it. If in two words Cache provides only functionality and control of the inner state but control of cache lvl and their collaboration will be provided by user (who will use this class in code)